### DISCUSSION TITLE
parseUnits large number bug

### DISCUSSION THREAD

**Discussion by mikemcdonald** [wevm/viem#896] [General]
Mentioned here: https://github.com/wagmi-dev/viem/discussions/490#discussioncomment-5907345, but opening a separate discussion for this bug since the `parseUnits` type was updated to a string already which was the original source of the discussion.

Example of a token on polygon with 0 decimals and the below balance:

```
const value: string = '12301000000000000020000';
parseUnits(value, 0);

SyntaxError: Cannot convert 1.2301e+22 to a BigInt
```

due to `Number()` https://github.com/wagmi-dev/viem/blob/main/src/utils/unit/parseUnits.ts#L12

Happy to submit a PR to fix - I see 3 routes to take:

1. BigInt round function to replace Math.round(Number(...))
2. Truncate decimals instead of rounding
3. Assert value decimals matches decimals provided which is what ethers.js does

**Comment by jxom** [comment]
Could probably just short-circuit with the bigint if the string does not have a fraction component right?

**Reply by mikemcdonald** [reply]
Yea that would solve this instance, but I think there would still be some edge cases from this line: https://github.com/wagmi-dev/viem/blob/c239cd7e62bd9cda0e8bd38777b08abb3764534a/src/utils/unit/parseUnits.ts#L21

**Reply by jxom** [reply]
Bigints don’t have a fraction component, so that condition will never be satisfied, right?

**Reply by mikemcdonald** [reply]
Ah yes you're right, originally read that wrong on how `rounded` was being used.

But going back to your first question, it's actually not relevant if the string does not have a fraction component. It's based on if the decimals provided is `0`. And then still have to decide what path to take from rounding, truncating, or throwing.
